ML-KEM/ML-DSA part 3: param array locate octet string

swenson opened the pull request:
Add OsslParam array locate octet string method.

Added back in from before commit deae632f6201f406fd3f5738feae78e4e61b295f

alex commented:
Is that really the only API in OpenSSL for getting the key bytes for an
ML-DSA key?

On Mon, Sep 8, 2025 at 8:34 AM Christopher Swenson ***@***.***>
wrote:

> *swenson* left a comment (sfackler/rust-openssl#2458)
> <https://github.com/sfackler/rust-openssl/pull/2458#issuecomment-3266089269>
>
> In the original PR, the .locate() function is used to take built EVP_PKEYs
> and get their public/private keys and seeds, for example.
> https://github.com/sfackler/rust-openssl/pull/2405/files#diff-af6f95b7f9e1a2c73e29a677932b06cb2b75ae3be91b7b691041b46c5829771fR181
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sfackler/rust-openssl/pull/2458#issuecomment-3266089269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAGBGRBGN233JGMUIZMVD3RVZXHAVCNFSM6AAAAACF372YZ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTENRWGA4DSMRWHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
All that is necessary for evil to succeed is for good people to do nothing.

alex commented:
Eh, we want the seeds (all non-seed serialization is dumb) so we're going to need something param based apparently.

alex commented on openssl/src/ossl_param.rs:
```diff
@@ -37,6 +39,78 @@ foreign_type_and_impl_send_sync! {
+
+/// A reference to a single `OSSL_PARAM` structure inside
+/// an `OsslParamArray`.
+pub(crate) struct OsslParam<'a>(&'a ffi::OSSL_PARAM);
```
This should probably named `OsslParamRef`

alex commented on openssl/src/ossl_param.rs:
```diff
@@ -37,6 +39,78 @@ foreign_type_and_impl_send_sync! {
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be valid and point to an `OSSL_PARAM` structure.
```
There's also a lifetime invariant.

alex commented on openssl/src/ossl_param.rs:
```diff
@@ -37,6 +39,78 @@ foreign_type_and_impl_send_sync! {
+        unsafe {
+            let mut bn: *mut ffi::BIGNUM = ptr::null_mut();
+            cvt(ffi::OSSL_PARAM_get_BN(self.as_ptr(), &mut bn))?;
+            Ok(&*(bn as *const BigNum))
```
This should use `from_ptr`

swenson replied:
I think I had `from_ptr` but switched away so I could annotate with a lifetime.

But I think we don't need this method at all. Let me see if I can simplify this.

alex reviewed (commented):
Can we add unit tests that verify these work with at least with a params created with the builder?

swenson commented:
@alex agreed re: seed formats are best.

I reworked this to combine `.locate()` and `.get_octet_string()` so that we don't need the extra struct at all. We only need octet strings for SLHDSA, MLDSA, and MLKEM, so I just deleted all of the string and BigNum stuff and it cleans up a lot.

And added a unit test.
